### ISSUE TITLE
Build an owned version of a message

### ISSUE THREAD

**Issue by bspeice** [capnproto/capnproto-rust#83]
Alright, I think I finally know enough to properly phrase this question.

Right now it seems that the only way to create a `TypedReader` is to build a message using the standard pattern, serialize, deserialize (which then owns the segments), and stuff into the `TypedReader`.

Is there a way to build the message using `OwnedSegments` to start with? Or get at the `OwnedSegments` from the constructed message? If we're using `HeapAllocator`, the builder presumably owns the data and it could be safe to move from builder into `TypedReader`.

EDIT: I suspect it has something to do with `Builder::get_segments_for_output()`, but I'm not sure how to get the `OutputSegments` -> `OwnedSegments`

**Comment by dwrensha** [comment]
Hm, interesting question! Maybe the way to go would be to implement `ReaderSegments` for `message::Builder`, like this:

```rust
impl <A> ReaderSegments for Builder<A> where A: Allocator {
    fn get_segment<'a>(&'a self, id: u32) -> Option<&'a [Word]> {
        self.get_segments_for_output().get(id as usize).map(|x| *x)
    }
}
```

Then you could create a `message::Reader` that wraps a `message::Builder`.

**Comment by bspeice** [comment]
Sorry it's taken me so long to respond. That indeed works! The result is a bit strange (because you need to wrap the `Builder` inside a `Reader` to get at the `ReaderSegments`) but here's an example:

```rust
#[test]
fn test_typed_builder() {

    let mut builder = message::Builder::new_default();

    {
        let mut ts_root = builder.init_root::<timestamp::Builder>();

        ts_root.set_seconds(14);

        ts_root.set_subsecond_nanos(0);
    }

    let reader = message::Reader::new(builder, message::ReaderOptions::new());

    let owned_message: TypedReader<message::Builder<message::HeapAllocator>, timestamp::Owned> = TypedReader::new(reader);

    assert_eq!(owned_message.get().unwrap().get_seconds(), 14);
}
```

**Comment by dwrensha** [comment]
Closed by #84.
